crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,753 @@
+        Self { method, params, block_id: None }
+    }
+
+    const fn with_block_id(mut self, block_id: BlockId) -> Self {
+        self.block_id = Some(block_id);
+        self
+    }
+
+    fn params_hash(&self) -> TransportResult<B256> {
+        // Merge the method + params and hash them.
+        let hash = serde_json::to_string(&self.params())
+            .map(|p| keccak256(format!("{}{}", self.method(), p).as_bytes()))
+            .map_err(RpcError::ser_err)?;
+
+        Ok(hash)
+    }
+
+    const fn method(&self) -> &'static str {
+        self.method
+    }
+
+    fn params(&self) -> Params {
+        self.params.clone()
+    }
+
+    /// Returns true if the BlockId has been set to a tag value such as "latest", "earliest", or
+    /// "pending".
+    const fn has_block_tag(&self) -> bool {
+        if let Some(block_id) = self.block_id {
+            match block_id {
+                BlockId::Hash(_) => return false,
+                BlockId::Number(BlockNumberOrTag::Number(_)) => return false,
+                _ => return true,
+            }
+        }
+        false
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct FsCacheEntry {
+    /// Hash of the request params
+    key: B256,
+    /// Serialized response to the request from which the hash was computed.
+    value: String,
+}
+/// Shareable cache.
+#[derive(Debug, Clone)]
+pub struct SharedCache {
+    inner: Arc<RwLock<LruMap<B256, String>>>,
+    max_items: u32,
+}
+
+impl SharedCache {
+    /// Instantiate a new shared cache.
+    pub fn new(max_items: u32) -> Self {
+        let inner = Arc::new(RwLock::new(LruMap::new(ByLength::new(max_items))));
+        Self { inner, max_items }
+    }
+
+    /// Maximum number of items that can be stored in the cache.
+    pub const fn max_items(&self) -> u32 {
+        self.max_items
+    }
+
+    /// Puts a value into the cache, and returns the old value if it existed.
+    pub fn put(&self, key: B256, value: String) -> TransportResult<bool> {
+        Ok(self.inner.write().insert(key, value))
+    }
+
+    /// Gets a value from the cache, if it exists.
+    pub fn get(&self, key: &B256) -> Option<String> {
+        // Need to acquire a write guard to change the order of keys in LRU cache.
+        self.inner.write().get(key).cloned()
+    }
+
+    /// Get deserialized value from the cache.
+    pub fn get_deserialized<T>(&self, key: &B256) -> TransportResult<Option<T>>
+    where
+        T: for<'de> Deserialize<'de>,
+    {
+        let Some(cached) = self.get(key) else { return Ok(None) };
+        let result = serde_json::from_str(&cached).map_err(TransportErrorKind::custom)?;
+        Ok(Some(result))
+    }
+
+    /// Saves the cache to a file specified by the path.
+    /// If the files does not exist, it creates one.
+    /// If the file exists, it overwrites it.
+    pub fn save_cache(&self, path: PathBuf) -> TransportResult<()> {
+        let cloned = self.inner.clone();
+        let cache = cloned.read();
```

Comment:
this is just read lock with extra steps,
what meant here is that should just clone the guarded value if we clone the values later anyway